metapool/metapool.py

```diff
@@ -2075,3 +2077,43 @@ def validate_plate_df(
             "The following sample names are duplicated %s"
             % ", ".join(sorted(duplicated_samples))
         )
+
+
+def generate_override_cycles_value(runinfo_xml_path, adapter_length):
+    if adapter_length < 0:
+        raise ValueError("Adapter-length cannot be less than zero")
+
+    tree = ET.parse(runinfo_xml_path)
+    reads = tree.getroot().findall('.Run/Reads/Read')
+    results = []
+    for read in reads:
+        result = {'number': int(read.get('Number')),
+                  'num_cycles': int(read.get('NumCycles'))}
+        result['is_indexed'] = True if read.get('IsIndexedRead') == 'Y' \
+            else False
+        results.append(result)
+
+    if len(results) == 0:
+        raise ValueError("Reads information could not be found in "
+                         f"'{runinfo_xml_path}'")
+
+    codes = []
+    # results should be in sorted order in XML file but this ensures that
+    # we process the elements in the correct order.
+    for read in sorted(results, key=itemgetter('number')):
```

Comment:
We loop through the reads to make results, and then loop through results to build up the codes.  Just so I'm clear, the reason it is necessary to loop over the same stuff twice is to ensure it is in order, because the order matters?

Reply:
Most definitely. I've never seen the reads appear out of order within a file but there's no guarantee. The ordering of each 'code' (for lack of a better word) has to match the order of the read in the BCL files.